nexus/mgs-updates/src/driver.rs

```diff
@@ -0,0 +1,586 @@
+                    // inactive slot".
+                    firmware_slot: 0,
+                    update_id,
+                };
+
+                (sp_update, Box::new(ReconfiguratorSpUpdater {}))
+            }
+        };
+
+        let baseboard_id = baseboard_id.clone();
+        let status_updater = UpdateAttemptStatusUpdater {
+            tx: self.status_tx.clone(),
+            baseboard_id: baseboard_id.clone(),
+        };
+        let artifacts = self.artifacts.clone();
+        let mgs_rx = self.mgs_rx.clone();
+        let request = internal_request.request.clone();
+        let future = async move {
+            let result = apply_update(
+                artifacts,
+                &sp_update,
+                &*updater,
+                mgs_rx,
+                &request,
+                status_updater,
+            )
+            .await;
+            UpdateAttemptResult { baseboard_id: request.baseboard_id, result }
+        }
+        .boxed();
+
+        let nattempts_done = internal_request.nattempts_done;
+        let in_progress = InProgressUpdate {
+            log: log.clone(),
+            time_started: chrono::Utc::now(),
+            instant_started: Instant::now(),
+            internal_request,
+        };
+
+        // Keep track of the work.
+        self.futures.push(future);
+
+        // Update status.
+        self.status_tx.send_modify(|driver_status| {
+            driver_status.in_progress.insert(
+                baseboard_id.clone(),
+                InProgressUpdateStatus {
+                    time_started: in_progress.time_started,
+                    instant_started: in_progress.instant_started,
+                    status: UpdateAttemptStatus::NotStarted,
+                    nattempts_done,
+                },
+            );
+        });
+
+        // Update our bookkeeping.
+        assert!(
+            self.in_progress
+                .insert(baseboard_id.clone(), in_progress)
+                .is_none()
+        );
+    }
+
+    /// Invoked when an in-progress update attempt has completed.
+    fn on_attempt_done(&mut self, result: UpdateAttemptResult) {
+        // Load the in-progress state and generate a CompletedAttempt status for
+        // this attempt.
+        let in_progress = self
+            .in_progress
+            .remove(&result.baseboard_id)
+            .expect("in-progress record for attempt that just completed");
+        let nattempts_done = in_progress.internal_request.nattempts_done + 1;
+        let completed = CompletedAttempt {
+            time_started: in_progress.time_started,
+            time_done: chrono::Utc::now(),
+            elapsed: in_progress.instant_started.elapsed(),
+            request: in_progress.internal_request.request.clone(),
+            result: match &result.result {
+                Ok(success) => Ok(success.clone()),
+                Err(error) => Err(InlineErrorChain::new(error).to_string()),
```

Comment:
Nit - I don't think we have to clone the success case here since we have an owned `result`

```suggestion
            result: result
                .result
                .map_err(|error| InlineErrorChain::new(&error).to_string()),
```

Reply:
Done.